\documentclass{article}
\usepackage[T1]{fontenc}
\usepackage[utf8]{inputenc}
\usepackage{charter} 
\usepackage[charter]{mathdesign}  
\usepackage[table]{xcolor}   % For table rows coloring
\usepackage{listings}        % To show when calling multiGWAS from console

\usepackage{amsmath}         % For writhing p-values
\newcommand{\mathA}[1]{{\operatorname{#1}}}
\newcommand{\mathB}[1]{{\operatorname{\mathit{#1}}}}


\usepackage[active]{srcltx}
\usepackage{textcomp}
\usepackage{graphicx}
\usepackage{authblk}
\usepackage{lineno}
\usepackage{subfigure}
\usepackage{multicol}
\usepackage{float}

% debe ser en formato APA pero en overleaf es sin parentesis después reviso en local

%----Added by LuisG. References in a file exported from Mendeley---
\usepackage{biblatex}
\addbibresource{multiGWAS.bib}
%---------------------------------------------



\makeatother
\linenumbers
\begin{document}




\title{MultiGWAS: A tool for GWAS analysis on tetraploid organisms by integrating the results of four GWAS software}

\author[1]{L. Garreta}

\author[1]{I. Cer\'{o}n-Souza}
\author[2]{M.R. Palacio}
\author[1]{P.H. Reyes-Herrera}

\affil[1]{Corporaci\'{o}n Colombiana de Investigaci\'{o}n Agropecuaria (AGROSAVIA), CI Tibaitat\'{a},  Kil\'{o}metro 14, V\'{i}a a Mosquera, 250047, Colombia}

\affil[2]{Corporaci\'{o}n Colombiana de Investigaci\'{o}n Agropecuaria (AGROSAVIA), CI El Mira, Kil\'{o}metro 38, V\'{i}a Tumaco Pasto, Colombia}


\maketitle

%%226/250
%IAquí hice unos cambios pequeños para ampliar el espectro del estudio a cualquier planta tetraploide ya que esta revista está enfocada en herramientas para estudios evolutivos de cualquier organismo. Att. Ivania
\begin{abstract}
\textbf{Summary:} The Genome-Wide Association Studies (GWAS) are essential to determine the association between genetic variants across individuals. One way to support the results is by using different tools to validate the reproducibility of the associations. Currently, software for GWAS in diploids is well-established but for polyploids species is scarce. Each GWAS software has its characteristics, which can cost time and effort to use them successfully. Here, we present MultiGWAS, a tool to do GWAS analysis in tetraploid organisms by executing in parallel and integrating the results from four existing GWAS software: two available for polyploids (GWASpoly and SHEsis) and two frequently used for diploids (PLINK and TASSEL). The tool deals with all the elements of the GWAS process in the four software, including (1) the use of different control quality filters for the genomic data, (2) the execution of two GWAS models, the full model with control for population structure and individual relatedness and the Naive model without any control. The summary report generated by MultiGWAS provides the user with tables and plots describing intuitively the significant association found by both each one and across four software, which helps users to check for false-positive or false-negative results.\\
%PAULA: no se si explicamos un poco mejor el output acá?
\textbf{Contact:} phreyes@agrosavia.co\\

\textbf{Keywords: GWAS, tetraploids, SNPs,XXX}

\end{abstract}
\maketitle 

\section{Introduction}

%Ya está modificada la intro para incluir diferentes especies de plantas poliploides. Faltan las citas. Estoy haciendo una lista de referencias que hay que leer con calma. Especialmente reviews. Intentaré en los próximos días ampliarhttps://www.overleaf.com/project/5e8b8de6ae23ed0001a9a14f este primer párrafo conforme vaya leyendo las referencias que bajé. Pero el sentido general creo que ya quedó establecido. Att. Ivania 

The GWAS (Genome-Wide Association Study) is used to identify which variants through the whole genome of a large number of individuals are associated with a specific trait (CITES). This methodology started with humans and several model plants, such as rice, maize, and $Arabidopsis$ \cite{tian2011genome, korte2013advantages}.  Because of the advances in the next-gen sequencing technology and the decreasing of the sequencing cost in recent years, there is an increase in genome sequences in non-model organisms at a faster rate \cite{ekblom2011applications, ellegren2014genome}. Therefore, several research projects pursue, for the first time, the genetic bases of an ecological or economic phenotypic variation through GWAS analysis for non-model wild plants and crops that often are polyploids \cite{ekblom2011applications, santure2018wild}. 

The GWAS for polyploid species has three related challenges. First, as all GWAS analyses, we should replicate the study as a reliable method to validate the results and recognize real associations. This replication involves finding the same associations either in an independent population sample under the same software or using an independent technology using the same population sample \cite{De2014,Pearson2008}. Both approaches involved the use of different parameters, models, or conditions, to test how consistent the results are. 

Second, although there are many GWAS software available to repeat the analysis under different conditions \cite{Gumpinger2018}, most of them are designed exclusively for the diploid data matrix \cite{Bourke2018}. Therefore, it is often necessary to "diploidizing" the polyploid genomic data in order to replicate the analysis. 

Third, the performance of different GWAS software could affect the results. However, there are few tools focused in the integration of the GWAS software outputs to make comparisons under different parameters and conditions. For example, comparing four packages for diploid species (i.e., PLINK, TASSEL, GAPIT, and FaST-LMM) using different sample sizes for plant data found that the threshold $P value$ for SNP significance change for each package \cite{Yan2019}. It means that well-ranked SNPs from one package can be ranked differently in another, causing difficulty in selecting the most credible associations when results from each package are analyzed separately. Besides, the software iPAT facilitates the use of three popular command-line GWAS packages GAPIT, PLINK, and FarmCPU \cite{Zhang2018}. However, results from the execution of each package are separated. Therefore, the problem of interpreting and selecting the best associations persists. In comparison, the easyGWAS cloud platform, perform, share, and compare the results of GWAS \cite{Grimm2017}. This platform offers two types of analysis: the first is an intersection analysis that searches associations that were found significant in more than one dataset. The second is a meta-analysis that searches associations mutually supported by several datasets. Both types are based on different datasets with the same GWAS parameters to confirm or search for new associations. The limitation of this approach is when the data set is unique or has a small sample of individuals. 

To solve all the three challenges above, we developed the MultiGWAS tool that performs GWAS analyses for tetraploid species using four software in parallel. Our tool include GWASpoly \cite{Rosyara2016} and the SHEsis tool \cite{Shen2016} that accept polyploid genomic data, and PLINK \cite{Purcell2007} and TASSEL \cite{Bradbury2007} with the use of a "diploidized" genomic matrix. The tool deals with preprocessing data, running four GWAS tools in parallel, and create reports that compare the results and help the user decide more intuitively the true or false associations.


\section{Method}

The MultiGWAS tool has three main steps: the adjustment state, the multi analysis stage, and the integration step (Fig. 1). In the first stage, we process the configuration file. Then the genotype and phenotype are cleaned and filtered and  we "diploidize" the genomic data. In the second stage, each GWAS tool runs in parallel. In the last stage, after the output files scanning, a summary of results is generated in a report that contains score tables, Venn diagrams, SNP profiles, and Manhattan plots. 

\begin{figure}
\includegraphics[width=12cm]{images/multi-gwas-flowchart-horizontal-stages-config.pdf}
\caption{MultiGWAS flowchart has three stages: adjustment, multi analysis, and integration.  }
\end{figure}

\subsection{Adjustment stage} 

MultiGWAS takes as input a configuration file where the user specifies the genomics data along with the parameters that will be used by the four tools. The configuration file includes the following parameters:

%genotype/phenotype filenames, genome-wide significance threshold, multiple testing correction methods, GWAS model, number of associations to be reported, and TRUE or FALSE whether to use quality control (QC) filters or not. 

 
\subsubsection{Genotype and phenotype files}
The path and filename for both files. The files are text files where the fields are separated by commas.

\subsubsection{GWAS model}
As MultiGWAS implements two types of GWAS analysis: (1) naive and (2) full. The naive model without any control for structure or relatedness between samples (without any additional parameters). Whereas the full model  takes into account for population structure (Q) and relatedness (K) to prevent false associations. Q and K  are obtained inside MultiGWAS by .... (default full or naivE?XXXX)


\subsubsection{Significance}
threshold .. (default XXXX)

\subsubsection{Correction}
Bonferroni .....

\subsubsection{Quality control filters}
TRUE or FALSE parameter to define if the QC filters are used. The filters considered at this stage are minor allele frequency (default XX), individual missing rate (default XX), SNP missing rate (default XX), Hardy\-Weinberg threshold (default XX).
\subsubsection{N associations}
N is the maximum number of associations that will be reported by each tool (default XXX)


Once the configurationf file is processed then the genomic data is preprocessed by selecting individuals present in both genotype and phenotype and excluding individuals and SNPs that have poor quality. Moreover, the format   "ACGT" suitable for the polyploid software GWASpoly and SHEsis, is "diploidized" for PLINK and TASSEL. The homozygous tetraploid genotypes are converted to diploid thus: (e.g.,AAAA\textrightarrow AA, CCCC\textrightarrow CC, GGGG\textrightarrow GG,
TTTT\textrightarrow TT). Moreover, for tetraploid heterozygous genotypes, the conversion depends on the reference and alternate alleles calculated for each position (e.g. AAAT\textrightarrow AT,
... ,CCCG\textrightarrow CG). After this process, MultiGWAS transform the genomic data to the formats required for each software.


\subsection{Multi analysis stage} 
MultiGWAS runs in parallel using two types of statistical models specified in the parameters file, the Full model (Q+K) and Naive (i.e., without any control) \cite{Sharma2018}. The Full model (Q+K) controls for both population structure and individual relatedness. For population structure, MultiGWAS uses the Principal Component Analysis (PCA) and takes the top ten PC as covariates. For relatedness, the tool uses kinship matrices that TASSEL and GWASpoly calculated separately, and for PLINK and SHEsis depends on the King software \cite{Manichaikul2010}. 

%\subsubsection{Tools}
We have selected four GWAS software tools to be integrated in our multiGWAS tool, two designed specifically for polyploid species as many important crops are polyploids: GWASpoly \cite{Rosyara2016} and SHEsis \cite{Yong2006}, and another two designed for diploids species and extensively used in humans and plants: PLINK \cite{Purcell2007,Chang2015} and TASSEL \cite{Bradbury2007}, respectively. 

As MultiGWAS implements two types of GWAS analysis, naive and full, each tool is called in two different ways. %The naive without any additional parameter, but the full with two parameters that take into account for population structure (Q) and relatedness (K) to prevent false associations.

\subsubsection{GWASpoly}

GWASpoly is a recent R package designed for GWAS in polyploid species that has been used in several studies in plants \cite{Berdugo2017,Ferrao2018,Sharma2018,Yuan2019}. It is based on the Q+K linear mixed model with biallelic SNPs that accounts for population structure and relatedness. In addition, to calculate the SNP effect for each genotypic class, GWASpoly provides a general gene action model along with four additional models: additive, simplex dominant, and duplex dominant. 

MultiGWAS is using GWASpoly version 1.3. The population structure and relatedness, used in the full model, are estimated using the first five principal components and the kinship matrix, respectively, both calculated with the algorithms built in GWASpoly. For both, naive and full models, all gene action models are tested for detecting associations.


\subsubsection{SHEsis}

SHEsis is another program designed for polyploid species that includes single locus association analysis, among others. It is based on a linear regresion model, and it has been used in some studies of animals and humans \cite{Qiao2015,Meng2019}. 

MultiGWAS is using the version 1.0 which does not take account for population structure or relatedness, however MultiGWAS externally estimates relatedness for SHEsis by excluding individuals with cryptic first-degree relatedness using the algorithm implemented in PLINK 2.0 (see below).


\subsubsection{PLINK}

PLINK is one of the most extensively used programs for GWAS in diploids species. It was developed for humans but it is applicable to any species \cite{Power2016}. PLINK includes a range of analysis, including univariate GWAS using two-sample tests and linear regression models.

MultiGWAS is using two versions of PLINK: 1.9 and 2.0. Linear regression from PLINK 1.9 is used to achieve both types of analysis, naive and full. For the full analysis, population structure is estimated using the first five principal components calculated with the PLINK 1.9 built in algorithm. But relatedness is estimated from the kinship coefficients calculated with the PLINK 2.0 built in algorithm, removing the close relatives or individuals with first-degree relatedness.

\subsubsection{TASSEL}

TASSEL is another common GWAS program based on the Java software. It was developed for maize and it has been used in several studies in plants \cite{Alvarez2017,Zhang2018}, but like PLINK, it is applicable to any species. For association analysis, TASSEL includes the general lineal model (GLM) and mixed linear model (MLM) that accounts for population structure and relatedness.

MultiGWAS is using TASSEL 5.0, with naive GWAS achieved by the GLM, and full GWAS achieved by the MLM with two parameters: one for population structure, using the first five principal components, and another for relatedness, using the kinship matrix with centered IBS method, both calculated with built in the TASSEL built in algorithms.


\subsection{Integration stage.} The outputs resulting from the four software are scanned and processed to identify both significant and best-ranked associations. Based on the specification of the configuration file about the correction method and the significant level, MultiGWAS corrects the p-values and calculates the threshold value for each associated marker. The calculation uses the number of valid genotype calls (i.e., non-missing genotypes, phenotypes, and covariates).  Then, MultiGWAS summarize the results in tables, Venn diagrams, SNP profiles, and  Manhattan plots.

\section{Results}
% Preview source code from paragraph 27 to 29

Although most of the GWAS packages used by MultiGWAS are based on a linear regression approaches, they often produce dissimilar association results for the same input. For example, computed \emph{p-values }for the same set of SNPs are different between packages; SNPs with significant \emph{p-values} for one package may be not significant for the others; or well-ranked SNPs in one package may be ranked differently in another. To alleviate these difficulties, MultiGWAS produces four types of outputs using different graphics and tabular views, including: score tables, Venn diagrams, Manhattan and Q-Q plots, and SNP profiles. These outputs are intended to help users to compare, select, and interpret the set of possible SNPs associated with a trait of interest. 

As an example of the functionality of the tool, here we show the results of running MultiGWAS tool in the genomic data from a tetraploid potato diversity panel, genotyped and phenotyped as part of the USDA-NIFA Solanaceae Coordinated Agricultural Project (SolCAP) \cite{Hirsch2013}. The reports include: significant SNPs, best-ranked SNPs, profile SNPs, and visualization of associations. First, the best-ranked SNPs (Figure \ref{fig:-View-Shared-SNPs}.a), where the SNP c2\_45606 was evaluated with a high score by the four packages, but other SNPs were also ranked with high scores by almost two packages. Second, the significant SNPs (Figure \ref{fig:-View-Shared-SNPs}.b), where the two polyploid software, GWASpoly and SHEsis, found as significant three SNPs, c1\_8019, c2\_25471, and c2\_45606, of which the c1\_8019 was also the most significant association found in the same potato dataset analyzed by Rosyara et al. (2016). Third, the SNP profiles (Figure \ref{fig:SNP-profiles}), where for each significant association, a heat map figure is generated to summarize the genotype associated with a trait for each individual. And fourth, the visualization of associations (Figure \ref{fig:view-qqmanhattan}), where for each package, a Manhattan and QQ plots are generated using special marks to help to identify significative, best-ranked, and shared SNPs (found by more than one tool). 

The complete report from MultiGWAS for the naive and full model is in the Supplementary information (\url{https://github.com/agrosavia-bioinformatic/multiGWAS}) 



\subsection{Visualization of shared SNPs}
GWAS packages compute \emph{p-value }as a measure of association between each individual SNP and the trait of interest. The SNPs are considered statistically significant, and so possible true associations, when their \emph{p-value }falls below a predefined significance level, usually 0.01 or 0.05. But, most GWAS packages compute differently both \emph{p-values }and significance levels, if the former are too high or the latter are too low, it could result in non-significant SNPs. Consequently, as it is important to know the significant SNPs, it is equally important to know the best-ranked SNPs closer to being statistically significant, as they may represent important associations to consider for posterior analysis (e.g. false negatives).

MultiGWAS provides tabular and graphic views to report in an integrated way both the best-ranked and significant SNPs identified by the four GWAS packages (see Figure \ref{fig:-View-Shared-SNPs}). Both \emph{p-values} and significance levels have been scaled as $-log_{10}(\mathB{p-value})$ to give high scores to the best statistically evaluated SNPs. First, the best-ranked SNPs correspond to the top-scored \emph{N} SNPs, wheter these SNPs were assesed significant or not by the package reporting them, and with\emph{ N} defined by the user in the configuration file. These SNPs are shown both in a SNPs table (Figure \ref{fig:-View-Shared-SNPs}.a) and in a Venn diagram (Figure \ref{fig:-View-Shared-SNPs}.b). The table lists them by package and sorts by decreasing score, whereas the Venn diagram shows them emphasizing if these ones were best-ranked either in a single package or in several at once (shared). And second, the significant SNPs correspond to the ones assesed statistically significant by each package, they are shown in a Venn diagram (Figure \ref{fig:-View-Shared-SNPs}.b), and they are also shown in the SNPs table, marked with significance TRUE and score greater than threshold, columns SGN, SCR, and THR, respectively in the table of the Figure\ref{fig:-View-Shared-SNPs}.a.

% Figure SNPs table and Venn diagrams
\begin{figure}[H]
{\fboxrule 0.2pt\fbox{\begin{minipage}[t][1\totalheight][b]{0.46\columnwidth}%
\renewcommand{\arraystretch}{1.5}
\setlength{\tabcolsep}{0.2em}
\rowcolors{2}{gray!25}{white}
\tiny
\begin{center}
\begin{tabular}{cccccccc}
\hline 
PKG & MDL & CHR & POS & SNP & SCR & THR & SGN\tabularnewline
\hline 
GWASpoly & Full & 10 & 48863165 & c1\_8019 & 4.78 & 4.25 & TRUE\tabularnewline
GWASpoly & Full & 10 & 48808404 & c2\_25471 & 4.57 & 4.27 & TRUE\tabularnewline
GWASpoly & Full & 10 & 48203431 & c2\_45611 & 4.36 & 4.27 & TRUE\tabularnewline
GWASpoly & Full & 10 & 48218826 & c2\_45606 & 4.68 & 4.5 & TRUE\tabularnewline
GWASpoly & Full & 4 & 71591813 & c2\_10692 & 2.93 & 4.25 & FALSE\tabularnewline
GWASpoly & Full & 4 & 71592285 & c2\_10687 & 2.93 & 4.25 & FALSE\tabularnewline
GWASpoly & Full & 4 & 71827521 & c2\_10614 & 2.93 & 4.25 & FALSE\tabularnewline
PLINK & Full & 10 & 67293176 & c1\_16001 & 1.7693 & 3.2601 & FALSE\tabularnewline
PLINK & Full & 10 & 77351069 & c1\_329 & 1.1795 & 3.301 & FALSE\tabularnewline
PLINK & Full & 11 & 51404231 & c2\_29435 & 1.1188 & 3.2553 & FALSE\tabularnewline
PLINK & Full & 10 & 69323144 & c2\_45611 & 1.0229 & 3.2553 & FALSE\tabularnewline
PLINK & Full & 2 & 41814861 & c2\_16350 & 0.9598 & 3.301 & FALSE\tabularnewline
PLINK & Full & 10 & 69311500 & c2\_45606 & 0.8489 & 3.2923 & FALSE\tabularnewline
PLINK & Full & 10 & 69809843 & c1\_16351 & 0.6131 & 3.2833 & FALSE\tabularnewline
SHEsis & Full & 2 & 13697423 & c1\_8019 & 9.4711 & 3.301 & TRUE\tabularnewline
SHEsis & Full & 1 & 30837971 & c1\_13526 & 8.4501 & 3.2923 & TRUE\tabularnewline
SHEsis & Full & 5 & 46046095 & c2\_53380 & 8.2409 & 3.2601 & TRUE\tabularnewline
SHEsis & Full & 3 & 39255236 & c2\_25471 & 7.8241 & 3.2923 & TRUE\tabularnewline
SHEsis & Full & 5 & 49804489 & c2\_54811 & 6.9633 & 3.2695 & TRUE\tabularnewline
SHEsis & Full & 1 & 69809843 & c1\_16351 & 6.0247 & 3.2833 & TRUE\tabularnewline
SHEsis & Full & 4 & 69311500 & c2\_45606 & 5.9557 & 3.2923 & TRUE\tabularnewline
TASSEL & Full & 8 & 54838024 & c2\_16999 & 3.6076 & 3.8943 & FALSE\tabularnewline
TASSEL & Full & 8 & 54838005 & c2\_16998 & 3.483 & 3.8943 & FALSE\tabularnewline
TASSEL & Full & 1 & 71450400 & c2\_12125 & 2.4832 & 3.8943 & FALSE\tabularnewline
TASSEL & Full & 1 & 70474651 & c2\_17191 & 2.45 & 3.8943 & FALSE\tabularnewline
TASSEL & Full & 1 & 70472380 & c2\_17193 & 2.2893 & 3.8943 & FALSE\tabularnewline
TASSEL & Full & 10 & 47539878 & c1\_16001 & 2.9101 & 4.5512 & FALSE\tabularnewline
TASSEL & Full & 7 & 14924207 & c2\_4342 & 2.8793 & 4.5512 & FALSE\tabularnewline
\hline 
\end{tabular}
\par\end{center}
\begin{center}
\normalsize (a)
\par\end{center}%
\end{minipage}}}~~~~~~~~~~~%
\noindent\begin{minipage}[t]{1\columnwidth}%
{\fboxrule 0.2pt\fboxsep 2pt\fbox{\begin{minipage}[t][1\totalheight][c]{0.38\columnwidth}%
\begin{center}
\includegraphics[bb=0bp 60bp 432bp 590bp,clip,scale=0.25]{images/out-multiGWAS-vennDiagram-best}\\
(b)
\par\end{center}%
\end{minipage}}}

~

{\fboxrule 0.2pt\fboxsep 2pt\fbox{\begin{minipage}[t][1\totalheight][c]{0.38\columnwidth}%
\begin{center}
\includegraphics[bb=0bp 60bp 432bp 590bp,clip,scale=0.25]{images/out-multiGWAS-vennDiagram-significatives}\\
(c)
\par\end{center}%
\end{minipage}}}%
\end{minipage}
\begin{centering}
\par\end{centering}
\caption{\scriptsize \textbf{Visualization of shared SNPs.} Tabular and graphical views of the best-ranked and significant SNPs identified by the four packages. \textbf{(a)} Tabular view with detailed information of each SNPs, including: package name (PKG), GWAS model used (MDL), chromosome (CHR), position in the genome (POS), ID (SNP), score (SCR), threshold (THR), and significance flag (SGN), wether the SNP was evaluated statistically significant or not (score > theshold). \textbf{(b)} Venn diagram with the best-ranked SNPs, showing that one SNP was shared by the four packages (c2\_45606), other two only by the two polyploid packages GWASpoly and SHEsis (c1\_8019 and c2\_25471), and other one only by the two diploid packages PLINK and TASSEL (c1\_16001). \textbf{(c)} Venn diagram with the significant SNPs, showing that only three SNPs (c1\_8019, c2\_25471, and c2\_45606) were evaluated as significant by the two polyploid packages GWASpoly and SHEsis.\label{fig:-View-Shared-SNPs}}
\end{figure}

\subsection{Visualization of Associations }
MultiGWAS creates a two-dimensional representation, called SNP profile, to visualize each trait by individuals and genotypes as rows and columns, respectively (Figure \ref{fig:SNP-profiles}). At the left, the individuals are grouped in a dendrogram by their genotype. At the right, there is the name or ID of each individual. At the bottom, the genotypes are ordered from left to right, starting from the major to the minor allele (i.e., AAAA, AAAB, AABB, ABBB, BBBB). At the top, there is a description of the trait based on a histogram of frequency (top left) and by an assigned color for each numerical phenotype value using a grayscale (top right). Thus, each individual appears as a colored line by its phenotype value on its genotype column. For each column, there is a solid cyan line with the mean of each column and a broken cyan line that indicates how far the cell deviates from the mean.

% Figure Manhattan and QQ plots
\begin{figure}[H]
\noindent %
\noindent\begin{minipage}[t]{1\columnwidth}%
\begin{minipage}[t][1\totalheight][c]{0.5\columnwidth}%
\begin{center}
\includegraphics[scale=0.5]{images/out-multiGWAS-views-manhattan-GWASpoly}\\
(a)
\par\end{center}%
\end{minipage}%
\begin{minipage}[t][1\totalheight][c]{0.5\columnwidth}%
\begin{center}
\includegraphics[bb=-30bp 0bp 294bp 238bp,clip,scale=0.5]{images/out-multiGWAS-views-QQ-GWASpoly}\\
(b)
\par\end{center}%
\end{minipage}%
\end{minipage}

\caption{\scriptsize\textbf{MultiGWAS visualization of associations.} MultiGWAS adds special marks to the Manhattan and QQ plots to help identify different types of SNPs: (a) In Manhattan plots, significant SNPs are above a red line, best-ranked SNPs are above a blue line, and shared SNPs (See Figure \ref{fig:-View-Shared-SNPs}.b) are colored in green (b) In QQ plots, a red diagonal line indicates the expectation, so potential associations can be observed when the number of SNPs deviating from the diagonal is small, as in the case of monogenic traits, or when this number is somewhat higher, as in the case of truly polygenic traits. However, deviations for a high number of SNPs could reflect inflated \emph{p-values }owing to population structure or cryptic relatedness. \label{fig:view-qqmanhattan}}
\end{figure}

\subsection{Visualization of common significative SNPs }
MultiGWAS creates a two-dimensional representation, called SNP profile, to visualize each trait by individuals and genotypes as rows and columns, respectively (Figure \ref{fig:SNP-profiles}). At the left, the individuals are grouped in a dendrogram by their genotype. At the right, there is the name or ID of each individual. At the bottom, the genotypes are ordered from left to right, starting from the major to the minor allele (i.e., AAAA, AAAB, AABB, ABBB, BBBB). At the top, there is a description of the trait based on a histogram of frequency (top left) and by an assigned color for each numerical phenotype value using a grayscale (top right). Thus, each individual appears as a colored line by its phenotype value on its genotype column. For each column, there is a solid cyan line with the mean of each column and a broken cyan line that indicates how far the cell deviates from the mean.

Because each multiGWAS report shows one specific trait at a time, the histogram and color key will remain the same for all the best-ranked SNPs.

% Figure SNP profiles
\begin{figure}[H]
\noindent\begin{minipage}[t]{1\columnwidth}%
\begin{minipage}[t]{0.5\columnwidth}%
\begin{center}
\includegraphics[scale=0.3]{images/out-SNPProfile-c2_45606}\\
(a)
\par\end{center}%
\end{minipage}%
\begin{minipage}[t]{0.5\columnwidth}%
\begin{center}
\includegraphics[scale=0.3]{images/out-SNPProfile-c1_8019}\\
(b)
\par\end{center}%
\end{minipage}%
\end{minipage}

\caption{\textbf{\scriptsize{}SNP profiles. }{\scriptsize{}SNP profiles for two of the best-ranked significant SNPs shown in the figure \ref{fig:-View-Shared-SNPs}.b. (a) SNP c2\_45606 best-ranked by the four packages (central intersection of the Venn diagram Figure \ref{fig:-View-Shared-SNPs}.b) (b) SNP c1\_8019 best-ranked by the two tetraploid packages (Figure \ref{fig:-View-Shared-SNPs}.b), and also identified as significant by the same packages (at the bottom of the Figure \ref{fig:-View-Shared-SNPs}.a). \label{fig:SNP-profiles}}}
\end{figure}

\section{Availability and implementation:} 
The core of the MultiGWAS tool was developed in R and users can interact with the tool by either a command line interface (CLI) developed in R or a graphical user interface (GUI) developed in Java (Figure [fig:MultiGWAS-interaction]). Source code, examples, documentation and installation instructions are available at https://github.com/agrosavia-bioinformatics/multiGWAS. 

\subsection{Input parameters}
MutiGWAS uses as the only input a simple configuration text file where users set the values for the main parameters that drives the GWAS process. The input parameters include: the output folder where results will be written, input genotype/phenotype filenames, genome-wide significance threshold, method for multiple testing correction, GWAS model, number of associations to be reported, and TRUE or FALSE whether to use quality control filters or not. The filters are: minor allele frequency, individual missing rate, SNP missing rate, and Hardy-Weinberg threshold.

The configuration file can be created either using a general text editor or using the GUI application. In the first case, the file must have the structure shown in the Figure \ref{fig:MultiGWAS-interaction}.a, where parameter names and values are separated by colon, filenames are enclosed in quotation marks, and TRUE or FALSE indicates wheter filters are applied or not. In the second case, the user creates the config file in a simple and straightforward way using the input parameter view from the GUI application (Figure \ref{fig:MultiGWAS-interaction}.b) and clicking the ``Save'' button.


\subsection{Using the command line interface}

The execution of the CLI tool is simple, it only needs to open a linux console, change to the folder where the configuration file was created, and type the name of the executable tool followed by the filename of the configuration file, like this:

\begin{lstlisting}[language=bash,basicstyle={\small}]
multiGWAS Test01.config
\end{lstlisting}

Then, the tool starts the execution, showing information of the process in the console window, and when it finishes the results are saved to a new subfolder called \emph{``outgwas/reports}. Results include a full html report containing the different views described in the results section, along with the original graphics and summary tables created by MultiGWAS and used to create the html report. Additionally, results include the preprocessed tables of the main outputs generated by the four GWAS packages used by MultiGWAS.

\subsection{Using the graphical user interface}
The MultiGWAS GUI application can be executed either by running from a Linux console the \emph{jmultiGWAS} command or by clicking on the Java application file \emph{JMultiGWAS.jar} located in the \emph{``multiGWAS/sources''} subfolder. After it opens, it shows a main frame with four tabs at the top (Figure \ref{fig:MultiGWAS-interaction}b): ``Inputs'', ``Outputs'', ``Results'', and ``Files''. The ``Inputs'' tab shows the form to create the configuration file and run the application. The ``Outputs'' tab shows the messages from the running process after it starts the execution. The ``Results'' tab shows the full html report described above. And the ``Files'' tab shows an embbeded file browser pointing to the subfolder that contains the original files used in the html report and described above.

\begin{figure}[H]
\begin{centering}
\begin{minipage}[t][1\totalheight][c]{0.5\columnwidth}%
\begin{center}
\includegraphics[scale=0.5]{images/screenshot-configFile}\\
(a)
\par\end{center}%
\end{minipage}
\par\end{centering}
~

~
\begin{centering}
\noindent\begin{minipage}[t]{1\columnwidth}%
\begin{center}
 \includegraphics[scale=0.38]{images/screenshot-jmultiGWAS-inputsView}\\
(b)
\par\end{center}%
\end{minipage}
\par\end{centering}
\caption{\scriptsize\textbf{MultiGWAS inputs and interaction}. MultiGWAS uses as input a simple configuration text file and can be executed using either a command line interface script in R (CLI) or a graphic user interface application in Java (GUI). \textbf{(a) }An example of a configuration text file named \emph{``Test01.config'' }including the parameters that drive the GWAS process. It can be created using a general text editor or using the GUI application (see below) \textbf{(b) }Main view of the MultiGWAS GUI application (``Inputs'' view) where users can create the configuration file by setting values for input parameters. The GUI contains other three views: ``Outputs'' view shows the logs of the running process. ``Results'' view shows a report in html format with the tabular and graphics described in the results section. And, the ``Files'' view shows an embedded file manager pointing to the subfolder that contains the files created by MultiGWAS and used to create the report. \protect \\
\label{fig:MultiGWAS-interaction}}
\end{figure}

















\section{Discussion}
%PAULA: Tools diploides vs Tools poliploides en poliploides 
XXXXXXXXXXXXXXXXXXXXXXX

%Also, the two diploid software, PLINK, and TASSEL identify a new marker, the c1\_16001. 



\printbibliography
% La bibliografia la coloqué toda en un archivo aparte "multiGWAS.bib" creado desde Mendeley y voy a compartirlerle un repositorio de Mendeley para que coloquen la bibliografía allá de forma más fácil y solo se exporte el archivo ("multiGWAS.bib") y se suba acá a overleaf (Yo puedo hacer eso)

\end{document}

