<!DOCTYPE html>
<html>
    <head>
        <link rel="stylesheet" href="stylesheets/metro-bootstrap.css">
	<link rel="stylesheet" href="stylesheets/iconFont.min.css">
        <script src="javascripts/jquery.min.js"></script>
        <script src="javascripts/jquery-ui.js"></script>
        <script src="javascripts/metro.min.js"></script>
    </head>
<body class="metro">

<div class="navigation-bar">
  <div class="navigation-bar-content container">
    <a class="element" href="http://analysis.bio-x.cn/SHEsisMain.htm" title="Back to SHEsis main page">
    <span class="icon-home"></span>
	SHEsisPlus
     </a>
    <span class="element-divider"></span>
    <a class="element" href="#">
      Help
    </a>
    <span class="element-divider"></span>
    <a class="element" href="CiteUs.html">
      Cite Us 
    </a>   
    <span class="element-divider"></span>
    <a class="element" href="ContactUs.html">
      Contact Us 
    </a> 
    <span class="element-divider"></span>
    <a class="element place-right" href="https://github.com/celaoforever/SHEsis" title="Follow us on github">
      Follow us on github
      <span class="icon-github-2"> 
    </a>
 </div>
</div>
<div class="container"  style="padding:20px 50px">
<h1>
 <a href="/">
  <i class="icon-arrow-left-3 fg-darker smaller"></i>
 </a>
  Help
 <small class="on-right"></small>
</h1>
<div class="container" style="padding:50px 100px 100px 100px">
<div style="line-height:30px">
<ul >
<font size="4">
<li><a href="#intro"  style="color:black">Introduction</a></li>
<br>
<li><a href="#new"  style="color:black">What's new</a></li>
<br>
<li><a href="#brow"  style="color:black">Browser Support</a></li>
<br>
<li><a href="#input"  style="color:black">Input Format</a></li>
<ul>
<br>
<li><a href="#inputBinary"  style="color:black">Case/Control Data</a></li>
<li><a href="#inputQTL"  style="color:black">Quantitative Trait Data</a></li>
</ul>
<br>
<li><a href="#arg"  style="color:black">Arguments</a></li>
<ul>
<br>
<li><a href="#analysis"  style="color:black">Choose Analysis</a></li>
<li><a href="#dataset"  style="color:black">Specify The Dataset</a></li>
<li><a href="#hap"  style="color:black">Extra Options For Haplotype Analysis</a></li>
<li><a href="#ld"  style="color:black">Extra Options For Linkage Disequilibrium</a></li>
<li><a href="#multi"  style="color:black">Multiple Comparision</a></li>
<li><a href="#mail"  style="color:black">Send A Mail When Completed</a></li>
</ul>
<br>
<li><a href="#output"  style="color:black">Interpret Output</a></li>
<ul>
<br>
<li><a href="#outputBinary"  style="color:black">Case/Control Data</a></li>
<li><a href="#outputQTL"  style="color:black">Quantitative Trait Data</a></li>
</ul>
<br>
<li><a href="#ref"  style="color:black">References </a></li>
</font>
</ul>
<br>
<br>

<h2 id="intro">Introduction</h2>
<br>
<font size="4">
<div style="text-indent:2em;padding:0px 0px 0px 0px">
SHEsisPlus is a web-based platform for analysis of genetic association, Hardy-weinberg equilibrium, linkage disequilibrium and haplotype construction at multiallelic polymorphism loci, compatible for both diploid and polyploid species. The web-based version provides most of the functions. The stand-alone version is also available. You can download it from <a href=#>here</a>.
</div>
<br>
<h2 id="new"> What's new </h2>
<br>
<font size="4">
<div style="text-indent:2em;padding:0px 0px 0px 0px">
Compared to previous version of SHEsis, SHEsisPlus is compatitable for haploid, diploid and polyploid species. It can not only analyze case/control data, but also quantitative trait data. It provides various ways of P value adjustments, including Holm step-down, Sidak single-step, Sidak step-down, FDR and permutation tests. All these can be can be performed via the webui.
</div>
<br>
<h2 id="brow">Browser Support</h2>
<table class="table bordered" >
<tr align="center">
<td style="width:200px">
<img src="http://www.w3schools.com/images/compatible_chrome.gif" style="padding:10px 20px 10px 40px" title="Chrome">
</td>
<td style="width:200px">
<img src="http://www.w3schools.com/images/compatible_firefox.gif" style="padding:10px 20px 10px 10px" title="Firefox">
</td>
<td style="width:200px">
<img src="http://www.w3schools.com/images/compatible_safari.gif" style="padding:10px 20px 10px 10px" title="Safari">
</td>
<td style="width:200px">
<img src="http://www.w3schools.com/images/compatible_opera.gif" style="padding:10px 20px 10px 10px" title="Opera">
</td>
<td style="width:200px">
<img src="http://www.w3schools.com/images/compatible_ie.gif" style="padding:10px 20px 10px 10px" title="IE">
</td>
</tr>
<tr align="center">
<td>Supoorted</td><td>Supported</td><td>Supported</td><td>Supported</td><td>Required IE9 or higher</td>
</tr>
</table>
<div style="text-indent:2em"> 
SHEsisPlus supports all of the modern browers, including Chrome, Firefox, Safari, Opera and IE. If you are using IE, please use IE9 or higher. 
</div>

<br>
<h2 id="input"> Input format</h2>
<br>
<h3 id="inputBinary">1. Case/control data</h3>
<div class="grid fluid">
<div class="row">
<div class="span6" style="padding:10px 0px 0px 50px">
<div class="panel">
  <div class="panel-header">
    sample data for diploidy species
  </div>
  <div class="content" style="padding:0px 0px 0px 0px">
    <pre>
    id1  G A  C C  1 1  A1 A2
    id2  A A  T C  1 1  A2 A2
    id3  A A  T T  2 2  A3 A4
    id4  0 0  T T  3 3  A5 A3
    id5  G G  A A  2 3  A1 A2
    id6  A A  C A  0 0  A6 A7
   </pre>
  </div>
</div>
</div>

<div class="span6" style="padding:10px 0px 0px 50px">
<div class="panel">
  <div class="panel-header">
    sample data for triploidy species
  </div>
  <div class="content" style="padding:0px 0px 0px 0px">
    <pre>
    id1  A G A  T C C  1 1 1  AA T  TT
    id2  A A A  C T C  2 1 1  A  T  AA
    id3  G A A  C T T  3 2 2  TT A  T
    id4  0 0 0  A T T  3 3 3  AA T  AT
    id5  G G G  T A A  1 2 3  TA TT T
    id6  G A A  C C A  0 0 0  AA A  A
   </pre>
  </div>
</div>
</div>
</div>
</div>

<br>
<div style="text-indent:2em">
The first column is sample id. The following columns are genotypes. They should be deliminated by space, comma or tab. Adjacent tokens will be compressed and will be treated as a single token. Genotypes can be any string (e.g. 1,2,3,4, or A,T,G,C , or A1,A2,A3,A4, or anything else) except 0, which is the coding for missing genotypes. 
</div>
<br>
<div style="text-indent:2em">
The above shown is sample data for diploid and tripolid species. For diploid species, the columns correspond to: sample id, site1-allele1, site1-allele2, site2-allele1, site2-allele2, .... For triploid species, the columns should be: sample id, site1-allele1, site1-allele2, site1-allele3, site2-allele1, site2-allele2, site2-allele3, ...
</div>
<br>
<h3 id="inputQTL">2. Quantitative trait data</h3>
<div class="grid fluid">
<div class="row">
<div class="span6" style="padding:10px 0px 0px 50px">
<div class="panel">
  <div class="panel-header">
    sample data for diploid species
  </div>
  <div class="content" style="padding:0px 0px 0px 0px">
    <pre>
    id1  20.6  G A  C C  1 1  A1 A2
    id2  25.4  A A  T C  1 1  A2 A2
    id3  23.1  A A  T T  2 2  A3 A4
    id4  42.4  0 0  T T  3 3  A5 A3
    id5  11.0  G G  A A  2 3  A1 A2
    id6  5.5   A A  C A  0 0  A6 A7
   </pre>
  </div>
</div>
</div>

<div class="span6" style="padding:10px 0px 0px 50px">
<div class="panel">
  <div class="panel-header">
    sample data for triploid species
  </div>
  <div class="content" style="padding:0px 0px 0px 0px">
    <pre>
    id1  1.1  A G A  T C C  1 1 1  AA T  TT
    id2  3.2  A A A  C T C  2 1 1  A  T  AA
    id3  14   G A A  C T T  3 2 2  TT A  T
    id4  4.3  0 0 0  A T T  3 3 3  AA T  AT
    id5  24   G G G  T A A  1 2 3  TA TT T
    id6  4.49 G A A  C C A  0 0 0  AA A  A
   </pre>
  </div>
</div>
</div>
</div>
</div>
<br>
<div style="text-indent:2em">
The format for quantitative trait data is similar to that for case/control data except that the second column is the quantitative trait. The quantitative trait should be numeric.
</div>
<br>
<h2 id="arg">Arguments </h2>
<br>
<h3 id="analysis">1. Choose analysis </h3>
<div style="text-indent:2em">
There are 4 type of analysis available: <i>association test, Hardy-Weinberg Equilibrum analysis, haplotype analysis and linkage disequilibrium analysis</i>. You need to choose the kind of tests you'd like to perform by clicking the checkbox in front of the analysis name. You should choose at least one kind of analysis, and you can also choose all the four analysis at the same time.
</div>
<br>
<h3 id="dataset">2. Specify the dataset </h3>
<div style="text-indent:2em">
SHEsisPlus supports analysis of polyploid species. You should specify <i>ploidy</i> of your dataset. Currently, web-based SHEsisPlus supports up to dodecaploid (12 sets). If the ploidy of your data is higher than 12, please download the stand-alone version.  
</div>
<br>
<div style="text-indent:2em">
<i>Phenotype</i> should also be specified. If the phenotype of your data is binary (case/control), please select "Case/Control" for phenotype. Then you should input data for both cases and controls in the below textarea. If your data is with quantitative trait, please select "Quantitative Trait" instead. Then in your data, every individual should have a quantitave trait and the quantitative trait should be numeric. 
</div>
<br>
<div style="text-indent:2em">
You could enter your maker names into the <i>Marker name</i> textfiled. Please use space, tab or comma to deliminate it. For example, "rs11,rs22,rs33,rs44" when there are 4 sites in all. This is opntional. If it is left blank, or invalid input is given, marker names will default to site1, site2, site3, ... 
</div>
<br>

<h3 id="hap">3. Extra options for Haplotype analysis</h3>
<div style="text-indent:2em">
If <i>Haplotype analysis</i> is choosen, you could define the <i>lowest frequency threshold (LFT)</i> of analysis. The default value of LFT is 0.03. Every single haplotype with a frequency less than this value will be discarded in the association analysis. The range of LFT is [0,1). This step is optional. Invalid input will default the value to 0.03.
</div>
<br>
<div style="text-indent:2em">
Moreover, you can also use a subset of the markers to perform haplotype. To do this, enter selected sites in <i>Mask for hap-analysis</i> textfiled. It should be deliminated by space, tab or comma. For example, "1,1,0,1" to choose 1st, 2nd and 4th sites for haplotype analysis when there are 4 site in all. If no mask is given, all the sites will be used.
</div>
<br>
<h3 id="ld">4. Extra options for Linkage disequilibrium</h3>
<div style="text-indent:2em">
For case/control data, linkage disequilibrium analysis can be carried out either in cases, in controls or in both cases and controls. You can select it after the label <i>Calculate LD in</i>. For quantitative trait data, linkage disequilibrium will be carried out in all samples.
</div>
<br>
<h3 id="multi">5. Multiple comparision</h3>
<div style="text-indent:2em">
We provided the methods to adjust for multiple testing. Two types of methods are available. One is p-value based methods including Holm's step down procedure, Sidal single-step procedure, Sidak step-down procedure, Benjamini & Hocheberg step-up FDR control and Benjamini & Yekutieli step-up FDR control. The other is permutation test, which is a data-level method and is much slower. If you want to perform p-value based adjustment, clicking the checkbox <i>Sidak, FDR, Holm's correction</i> after the label <i>Multiple comparision</i>. If you want to perform permutation test, enter the number of permutations in the textfiled <i>Number of permutations</i>. Currently, SHEsisPlus webserver accepts number of permutations <10000.
</div>
<br>
<h3 id="mail">6. Send a mail when completed</h3>
<div style="text-indent:2em">
Sometimes the analysis takes a long time, especially when your dataset is very large, or permutation test is selected. Under such circumstance, you can provide your e-mail address in the textfiled <i>Your E-mail</i>. When the analysis is done, the link to the results will be sent to your mailbox. Additionally, you can also specify the name of the dataset, which will be included in the mail subject. 
</div>
<br>
<h2 id="output">Interpret output</h2>
<br>
<h3 id="outputBinary">1. Binary phenotype</h3>
<br>
<h4>a. Association test</h4>

<img src="images/help-assoc-binary.PNG" style="padding:10px 20px 10px 10px">
<br>
<div style="text-indent:2em">
This is an exmple of output for case/control association analysis. The fileds are:
</div>
<div class="container" style="padding:20px 200px 10px 50px">
<table style="width:auto;padding:50px 50px 50px 50px" class="table  bordered hovered " id="Allele_Association">
<tbody>
<tr style="font-weight:normal ">
<th class="text-left info"  style="width:150px">Header</th>
<th class="text-left info">Explanation</th>
</tr>
<tr>
<td>SNP</td><td>Names of SNPs. Your can specify the them in the textfiled <i>Marker names</i>. If you don't provide them, or your input is invalid, SNP names will default to site1, site2, site3 ...</td>
</tr>
<tr>
<td>Call rate</td><td>The percentage of individuals with nonmissing genotypes</td>
</tr>
<tr>
<td>Chi<sup>2</sup></td><td>χ<sup>2</sup> in Pearson's Chi square test</td>
</tr>
<tr>
<td>Pearson's p</td><td>p value calculated from Pearson's Chi square test</td>
</tr>
<tr>
<td>Fisher's p</td><td>p value calculated from Fisher's exact test</td>
</tr>
<tr>
<td>Permutation p</td><td>p value acquired from permutation test</td>
</tr>
<tr>
<td>OR [95% CI]</td><td>Odds ratio [95% confident interval]. Please note that this value is only presented when a site has two allele types.</td>
</tr>
<tr>
<td>Holm</td><td>Holm (1979) step-down adjusted p-values for strong control of the family-wise Type I error rate (FWER)</td>
</tr>
<tr>
<td>SidakSS</td><td>Sidak single-step adjusted p-values for strong control of the family-wise Type I error rate (FWER)</td>
</tr>
<tr>
<td>SidakSD</td><td>Sidak step-down adjusted p-values for strong control of the family-wise Type I error rate (FWER)</td>
</tr>
<tr>
<td>FDR_BH</td><td>adjusted p-values for the Benjamini & Hochberg (1995) step-up FDR controlling procedure</td>
</tr>
<tr>
<td>FDR_BY</td><td>adjusted p-values for the Benjamini & Yekutieli (2001) step-up FDR controlling procedure</td>
</tr>
<tr>
<td>Detail</td><td>The counts and frequencies of a specific genotype or alleles</td>
</tr>
</tbody>
</table>
</div>
<h4>b. Hardy-weinberg equilibrium test</h4>
<img src="images/help-hwe-binary.PNG" style="padding:10px 20px 10px 10px">
<br>
<div style="text-indent:2em">
The output is straightforward. Hardy-weinberg equilibrium is calculated in cases, in controls, and in both cases and controls. Both Pearson's chi square test and Fisher's exact test are performed. 
</div>
<br>
<h4>c. Haplotype analysis</h4>
<div style="text-indent:2em">
For web-based SHEsisPlus, the algorithm used for haplotype interfernce is expectation maximization (EM). Another algorithm based on <a href="http://www.ncbi.nlm.nih.gov/pubmed/18667059">SAT</a> is available for stand-alone version. The SAT-based algorithm is much more slower than EM algorithm.The output for haplotype analysis looks like: 
</div>
<img src="images/help-hap-binary.PNG" style="padding:10px 20px 10px 10px">
<br>
<div style="text-indent:2em">
 In this example, haplotypes with frequency <0.03 are discarded. 0.03 is the default value. You can change this value by modifying textfield <i>Lowest frequency for hap-analysis</i>. The fields in the table are also easy to understand. Apart from association test for every single haplotype, a global result is also given. This result shows if the haplotype distribution is different between cases and controls.
</div>
<br>
<h4>d. Linkage disequilibrium analysis</h4>
<img src="images/help_binary_D.jpg" style="width:40%;height:40%;padding:10px 20px 10px 10px">
<img src="images/help_binary_R2.jpg" style="width:40%;height:40%;padding:10px 20px 10px 10px">
<div style="text-indent:2em">
For linkage disequilibrium analysis, pair-wise D' and R<sup>2</sup> are calculated. The higher two loci are in linkage disequilibrium, the darker the color will be. 
</div>
<br>
<h3 id="outputQTL">1. Quntitative trait</h3>
<br>
<h4>a. Association test</h4>
<img src="images/help-assoc-qtl.PNG" style="padding:10px 20px 10px 10px">
<div style="text-indent:2em">
This is an exmple of output for case/control association analysis. The fileds are: 
</div>

<div class="container" style="padding:20px 200px 10px 50px">
<table style="width:auto;padding:50px 50px 50px 50px" class="table  bordered hovered " id="Allele_Association">
<tbody>
<tr style="font-weight:normal ">
<th  class="text-left info"  style="width:150px">Header</th><th  class="text-left info">Explanation</th>
</tr>
<tr>
<td>SNP</td><td>Names of SNPs. Your can specify the them in the textfiled <i>Marker names</i>. If you don't provide them, or your input is invalid, SNP names will default to site1, site2, site3 ...</td>
</tr>
<tr>
<td>Effect allele</td><td>Contributing allele. For biallelic species, the effect allele is the minor allele. For multiallelic species, the effect allele is the allele that gives the lowest p value.</td>
</tr>
<tr>
<td>Nonmissing</td><td>Number of non-missing individuals included in analysis</td>
</tr>
<tr>
<td>Beta</td><td>Regression coefficient</td>
</tr>
<tr>
<td>SE</td><td>Standard error</td>
</tr>
<tr>
<td>R<sup>2</sup></td><td>Regression r-squared</td>
</tr>
<tr>
<td>T</td><td>Wald test (based on t-distribtion)</td>
</tr>
<tr>
<td>p</td><td>Wald test asymptotic p-value</td>
</tr>
<tr>
<td>permutation p</td><td>p value acquired from permutation test</td>
</tr>
</tbody>
</table>
</div>
<div style="text-indent:2em">
The following fields have been described before. 
</div>
<br>
<h4>b. Hardy-weinberg equilibrium test</h4>
<img src="images/help-hwe-qtl.PNG" style="padding:10px 20px 10px 10px">
<div style="text-indent:2em">
For quantative trait, Hardy-weinberg equilibrium test in all samples are carried out. 
</div>
<br>
<h4>c. Haplotype analysis</h4>
<img src="images/help-hap-qtl.PNG" style="padding:10px 20px 10px 10px">
<div style="text-indent:2em">
 <i>Total count</i> is the count of this haplotype in the dataset. The other fileds have been described before 
</div>
<br>
<h4>d. Linkage disequilibrium analysis</h4>
<div style="text-indent:2em">
For quantitative trait, linkage disequilibirum is calculated in all samples. The results are similar to that of case/control data.
</div>
<br>
<h2 id="ref">References:</h2>
<br>
<div>
[1] Neigenfind J1, Gyetvai G, Basekow R, Diehl S, Achenbach U, Gebhardt C, Selbig J, Kersten B.Haplotype inference from unphased SNP data in heterozygous polyploids based on SAT. <i>BMC Genomics</i> 2008 Jul 30;9:356. doi: 10.1186/1471-2164-9-356.
</div>
<div>
[2] Purcell S1, Neale B, Todd-Brown K, Thomas L, Ferreira MA, Bender D, Maller J, Sklar P, de Bakker PI, Daly MJ, Sham PC.PLINK: a tool set for whole-genome association and population-based linkage analyses.<i> Am J Hum Genet</i> 2007 Sep;81(3):559-75. Epub 2007 Jul 25.
</div>
<div>
[3] Hedrick PW. Gametic disequilibrium measures: proceed with caution. <i>Genetics</i> 1987 Oct;117(2):331-41.
</div>


</div>
</font>
</div>
</div>
</div>
</body>
</html>
